solr/core/src/java/org/apache/solr/search/facet/FacetParser.java

```diff
@@ -138,21 +139,30 @@ public Object parseFacetOrStat(String key, Object o) throws SyntaxError {
     return parseFacetOrStat(key, type, args);
   }
 
+  public interface ParseHandler {
+    Object doParse(FacetParser<?> parent, String key, Object args) throws SyntaxError;
+  }
+
+  private static final Map<String, ParseHandler> REGISTERED_TYPES = new ConcurrentHashMap<>();
+
+  static {
+    ParseHandler fieldParser = (p, k, a) -> new FacetFieldParser(p, k).parse(a);
+    REGISTERED_TYPES.put("field", fieldParser);
+    REGISTERED_TYPES.put("terms", fieldParser);
+    REGISTERED_TYPES.put("query", (p, k, a) -> new FacetQueryParser(p, k).parse(a));
+    REGISTERED_TYPES.put("range", (p, k, a) -> new FacetRangeParser(p, k).parse(a));
+    REGISTERED_TYPES.put("heatmap", (p, k, a) -> new FacetHeatmap.Parser(p, k).parse(a));
+    REGISTERED_TYPES.put("func", (p, k, a) -> p.parseStat(k, a));
+  }
+
+  public static void registerParseHandler(String type, ParseHandler parseHandler) {
```

Comment:
Thanks for the https://issues.apache.org/jira/browse/SOLR-17477 comment w.r.t. how this is called!

Reply:
Not sure where to discuss it. But registering anything in static initializer is asking for trouble, imho. I'd rather hookup a component via solrconfig.xml (or a plugin??), which just calls `registerParserHandler()` for custom handlers.
Anyway it's a matter of taste, and should block this PR, unless someone propose to make  `FP.registerParserHandler` `protected`. WDYT?

Reply:
Yeah it's perhaps not the best way of hooking things up, especially as the whole process is already kicked off from solrconfig.xml where you declare a valueSourceParser. I was relying on the point at which that custom parser object is instantiated to run a static initialiser to also register it for json parsing too. But maybe it's safer to invert that and have the Solr code register the custom parser if it implements json parsing. I'll have another look and think about the alternatives.

My main reason for suggesting custom code uses a static block to register is when you have many hundreds of SolrCores in your CoreContainer and each one would be instantiated and call the register method, when really you might only need it done once. But now thinking if you run a mixed workload with different solrconfig and they're not homogenous, you don't want static registering maybe (each SolrCore is distinct)

Reply:
Although the 'standard set' of registered types in the static block above probably is appropriate as those are built-in and common to all SolrCores.

Reply:
`REGISTERED_TYPES` are shared across cores. It means:
- one core might inject a custom parser, it appears across all cores. If it hooks up some large state it remains in heap after core unloading (it might be called as a leak)
- `registerPH` might be called with one of the standard name bringing some surprises to other cores.

Reply:
Yes all good points, I'm going to change this to remove the statics and make it per-core which probably fits better with how it's hooked in too